include/openthread/dnssd_server.h

```diff
@@ -147,6 +147,30 @@ typedef enum
     OT_DNSSD_QUERY_TYPE_RESOLVE_HOST = 3, ///< Service type resolve hostname.
 } otDnssdQueryType;
 
+/**
+ * Includes the failure counters due to different reasons.
+ *
+ */
+typedef struct otUpstreamDnsFailureCounters
+{
+    uint32_t mTotalFailure; ///< Sum up of all failures.
```

Comment:
Actually the "no memory" error may not be that useful. Suggest just tracking total failure count?

We should move the upstream DNS code to use heap allocated entries rather fixed array definition of `UpstreamQueryTransaction mUpstreamQueryTransactions[kMaxConcurrentUpstreamQueries];`.

Reply:
I think if we have multiple failure cases, we should add a extra total counter. This should increase the code level backward compatibility.

Consider a code is written before adding a new error case, it will miss the error counter, and makes debug more difficult (since the error counters indicates that there are no errors)

When we have a total counter, then the users will be aware that "there is an error here, but the user's code is a bit too old to tell more details"

Reply:
> Actually the "no memory" error may not be that useful. Suggest just tracking total failure count?
> 
> We should move the upstream DNS code to use heap allocated entries rather fixed array definition of `UpstreamQueryTransaction mUpstreamQueryTransactions[kMaxConcurrentUpstreamQueries];`.

Makes sense, I would suggets using a build time switch to support both?

Reply:
Ok, I moved the totalFailures to otDnssdCounters, if we have other failure reasons in the furture need to be abstracted from total failures, we can add more counters.